### PR TITLE
Fix #5404 : Migrate onBackPressed

### LINKED ISSUE #5404: [Feature Request]: Migrate away from onBackPressed()
### Is your feature request related to a problem? Please describe.

API 33 (introduced in #5222) deprecated onBackPressed: https://developer.android.com/reference/android/app/Activity#onBackPressed().

#5402 helped reveal which activities need updating:
- [x] app/src/main/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivity
- [x] app/src/main/java/org/oppia/android/app/administratorcontrols/appversion/AppVersionActivity
- [x] app/src/main/java/org/oppia/android/app/administratorcontrols/learneranalytics/ProfileAndDeviceIdActivity
- [x] app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedActivity
- [x] app/src/main/java/org/oppia/android/app/devoptions/markstoriescompleted/MarkStoriesCompletedActivity
- [x] app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/MarkTopicsCompletedActivity
- [x] app/src/main/java/org/oppia/android/app/options/ReadingTextSizeActivityPresenter
- [x] app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter
- [x] app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonActivityPresenter
- [x] app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivity
- [x] app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivityPresenter
- [x] app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivityPresenter
- [x] app/src/main/java/org/oppia/android/app/walkthrough/end/WalkthroughFinalFragmentPresenter

### Describe the solution you'd like

As mentioned in https://developer.android.com/reference/android/app/Activity#onBackPressed(), the solution here seems to be to:
- No longer override ``onBackPressed`` directly.
- To use AndroidX (via ``androidx.activity.ComponentActivity#getOnBackPressedDispatcher()``) to register a backward-compatible back pressed handler.
- Migrate existing ``onBackPressed`` calls to the new dispatcher.

### Describe alternatives you've considered

The only alternative is not using AndroidX, but this isn't compatible with Oppia Android since the new deprecation started in SDK 33 (and the app's min supported SDK in 21).

### Additional context

Folks who work on this issue should ensure that back navigation behavior continues to work on changed activities & fragments, specifically:
- Using the back arrow or x button in the toolbar (if present) correctly navigates back to the previous activity.
- Using the system back button correctly navigates to the previous activity.
- If there's any special handling in the changed activity/fragment for tablets, navigation and split fragment displays should be verified on a tablet version of Android.

NB: The first work on this issue will require updating AndroidX core to a newer version.

### PR DESCRIPTION
**Pull request by Tejas-67**
## Explanation
Fix #5404
For devices running on API 33 or above, onBackPressedDispatcher doesn't work, hence we need to use onBackInvokeDispatcher.

In the app:manifest,
`<application android:enableOnBackInvokedCallback="true"/>`
has been added so that the onBackInvokedDispatcher works.

- [x]  app/src/main/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivity
- [x]  app/src/main/java/org/oppia/android/app/administratorcontrols/appversion/AppVersionActivity
- [x]  app/src/main/java/org/oppia/android/app/administratorcontrols/learneranalytics/ProfileAndDeviceIdActivity
- [x]  app/src/main/java/org/oppia/android/app/devoptions/markchapterscompleted/MarkChaptersCompletedActivity
- [x]  app/src/main/java/org/oppia/android/app/devoptions/markstoriescompleted/MarkStoriesCompletedActivity
- [x]  app/src/main/java/org/oppia/android/app/devoptions/marktopicscompleted/MarkTopicsCompletedActivity
- [x]  app/src/main/java/org/oppia/android/app/options/ReadingTextSizeActivityPresenter
- [x]  app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter
- [x]  app/src/main/java/org/oppia/android/app/resumelesson/ResumeLessonActivityPresenter
- [x]  app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivity
- [x]  app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivityPresenter
- [x]  app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivityPresenter
- [ ]  app/src/main/java/org/oppia/android/app/walkthrough/end/WalkthroughFinalFragmentPresenter

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by Tejas-67** [comment]
@adhiamboperes please check. If this is alright then I would proceed with the others as well.

File: app/src/main/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivity.kt [hunk 1]
```diff
@@ -86,6 +88,46 @@ class AdministratorControlsActivity :
       isProfileDeletionDialogVisible
     )
     title = resourceHandler.getStringInLocale(R.string.administrator_controls)
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+      onBackInvokedDispatcher.registerOnBackInvokedCallback(1000) {
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivity.kt]: This needs to be a constant, e.g. `OnBackInvokedDispatcher.PRIORITY_DEFAULT`

File: app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivityPresenter.kt [hunk 2]
```diff
@@ -43,6 +35,9 @@ class ProfileEditActivityPresenter @Inject constructor(
     }
   }
 
+  private fun getProfileEditActivity(): ProfileEditActivity? {
+    return activity as ProfileEditActivity?
+  }
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivityPresenter.kt]: Why is this needed? It does not appear to be in use.

File: app/src/main/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivity.kt [hunk 3]
```diff
@@ -86,6 +88,46 @@ class AdministratorControlsActivity :
...
+         * If it's instead ProfileEditFragment then profileListFragment should be inflated via
+         * handleOnBackPressed.
+         */
+        if (fragment is ProfileEditFragment) {
+          administratorControlsActivityPresenter.handleOnBackPressed()
+        } else {
+          onBackPressedDispatcher.onBackPressed()
+        }
+      }
+    } else {
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivity.kt]: I propose removing this bit of the onBackInvokedDispatcher.registerOnBackInvokedCallback() and remaining with the part that is under "else".

According to the documentation, https://developer.android.com/reference/android/app/Activity#onBackPressed(), `onBackPressedDispatcher.addCallback()` is backward compatible, eliminating the need to gate the logic with the sdk version.

File: app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivity.kt [hunk 4]
```diff
@@ -41,19 +43,33 @@ class ProfileEditActivity : InjectableAutoLocalizedAppCompatActivity() {
...
+      onBackInvokedDispatcher.registerOnBackInvokedCallback(1000) {
+        val isMultipane: Boolean = intent.extras!!.getBoolean(IS_MULTIPANE_EXTRA_KEY, false)
+        if (isMultipane) {
+          onBackPressedDispatcher.onBackPressed()
+        } else {
+          val intent = Intent(this, ProfileListActivity::class.java)
+          intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+          startActivity(intent)
+        }
+      }
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivity.kt]: Same comment regarding gating sdk

File: app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivity.kt [hunk 5]
```diff
@@ -41,19 +43,33 @@ class ProfileEditActivity : InjectableAutoLocalizedAppCompatActivity() {
...
+              val intent = Intent(this@ProfileEditActivity, ProfileListActivity::class.java)
+              intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+              startActivity(intent)
+            }
+          }
+        })
     }
+    (activityComponent as ActivityComponentImpl).inject(this)
+    profileEditActivityPresenter.handleOnCreate()
   }
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditActivity.kt]: We can improve readability by putting this in a function.
```suggestion
override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    ....
    handleBackPress()
  }

  private fun handleBackPress() {
    onBackPressedDispatcher.addCallback(
      this@AdministratorControlsActivity, object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
          val fragment =
            supportFragmentManager.findFragmentById(
              R.id.administrator_controls_fragment_multipane_placeholder
            )
          /*
           * If the current fragment is ProfileListFragment then the activity should end on back press.
           * If it's instead ProfileEditFragment then profileListFragment should be inflated via
           * handleOnBackPressed.
           */
          if (fragment is ProfileEditFragment) {
            administratorControlsActivityPresenter.handleOnBackPressed()
          } else {
            isEnabled = false
            onBackPressedDispatcher.onBackPressed()
            isEnabled = true
          }
        }
      })
  }
```

File: app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -14,6 +14,7 @@
     android:icon="@mipmap/launcher_icon"
     android:label="@string/app_name"
     android:supportsRtl="true"
+    android:enableOnBackInvokedCallback="true"
```
Comment (adhiamboperes) [inline app/src/main/AndroidManifest.xml]: This is not needed as far as I can tell.

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @Tejas-67!

The main comment is to use `onBackPressedDispatcher.addCallback()` for backward compatibility, but I have seen your comment in the PR description that this doesn't work above api 33. Could you please confirm that?

I would also encourage handling the back press in a function that you can call, that will remove the need to move code from the presenter to activity or fragment in some cases.

**Comment by Tejas-67** [comment]
hey @adhiamboperes, I have migrated the activities, but facing problem in the WalkthroughFinalFragmentPresenter. Even after setting up the onBackPressedDispatcher in the WalkthroughFinalActivity, I am unable to get the reference of the onBackPressedDispatcher in the Presenter. how can i fix this?

**Comment by adhiamboperes** [comment]
> hey @adhiamboperes, I have migrated the activities, but facing problem in the WalkthroughFinalFragmentPresenter. Even after setting up the onBackPressedDispatcher in the WalkthroughFinalActivity, I am unable to get the reference of the onBackPressedDispatcher in the Presenter. how can i fix this?

Could you please share a screenshot of your implementation so that I can understand your issue better?

**Comment by Vishwajith-Shettigar** [comment]
Hi @Tejas-67, could you please provide an update on your progress? If you are facing any blockers, let us know. Kindly attach all the relevant details and push the code that is not working, so we can have a clearer understanding of the issues.

**Comment by Tejas-67** [comment]
hey @Vishwajith-Shettigar ,  I am done with the WalkthroughFinalPresenter as well. But I am facing this error while pushing the code.

`buf_lint_check.sh: line 13: ../oppia-android-tools/buf-Darwin-x86_64: No such file or directory

Protobuf lint check issues found. Please fix them before pushing your code`

  Any idea on how to fix this?

**Comment by Vishwajith-Shettigar** [comment]
`Protobuf lint check issues found. Please fix them before pushing your code.`

Since I'm a Windows user, I never encountered this (this lint check doesn't support Windows, I assume), but I can see you may have made changes in protobuf files. If so, check for lint issues. But I don't have a clue about this (maybe you missed some commands while setting up oppia-android) `buf_lint_check.sh: line 13: ../oppia-android-tools/buf-Darwin-x86_64: No such file or directory.`

@adhiamboperes may help you!

**Comment by adhiamboperes** [comment]
@Tejas-67 Could you please confirm that you have the `android-tools` directory?

Screenshot 2024-07-04 at 23 56 13

If not, on your terminal, cd into the project root and run command `bash scripts/setup.sh`
You should be able to see inside the generated folder:

Screenshot 2024-07-04 at 23 59 56


Please let me know if this works, or revisit the wiki installation part for M1 Mac, and look at the instructions for using a Rossetta terminal.

**Comment by Tejas-67** [comment]
@adhiamboperes I have the oppia-android-tools folder already, but i have buf-Darwin-arm64 instead of buf-Darwin-x86_64 .

**Comment by adhiamboperes** [comment]
> @adhiamboperes I have the oppia-android-tools folder already, but i have buf-Darwin-arm64 instead of buf-Darwin-x86_64 .


Do you use intel Mac or M1/M2

**Comment by Tejas-67** [comment]
@adhiamboperes , M2

**Pull request closed without merging**